jetty-util/src/main/java/org/eclipse/jetty/util/thread/ReservedThreadExecutor.java

```diff
@@ -266,139 +265,144 @@ public String toString()
         return String.format("%s@%x{s=%d/%d,p=%d}",
             getClass().getSimpleName(),
             hashCode(),
-            _size.get(),
+            _count.getLo(),
             _capacity,
-            _pending.get());
+            _count.getHi());
     }
 
-    private class ReservedThread implements Runnable
+    private enum State
     {
-        private final SynchronousQueue<Runnable> _task = new SynchronousQueue<>();
-        private boolean _starting = true;
-
-        public boolean offer(Runnable task)
-        {
-            if (LOG.isDebugEnabled())
-                LOG.debug("{} offer {}", this, task);
-
-            try
-            {
-                _task.put(task);
-                return true;
-            }
-            catch (Throwable e)
-            {
-                LOG.ignore(e);
-                _size.getAndIncrement();
-                _stack.offerFirst(this);
-                return false;
-            }
-        }
+        PENDING,
+        RESERVED,
+        RUNNING,
+        IDLE,
+        STOPPED
+    }
 
-        public void stop()
-        {
-            offer(STOP);
-        }
+    private class ReservedThread implements Runnable
+    {
+        // The state and thread are kept only for dumping
+        private volatile State _state = State.PENDING;
+        private volatile Thread _thread;
 
         private Runnable reservedWait()
         {
             if (LOG.isDebugEnabled())
-                LOG.debug("{} waiting", this);
+                LOG.debug("{} waiting {}", this, ReservedThreadExecutor.this);
 
             while (true)
             {
                 try
                 {
-                    Runnable task = _idleTime <= 0 ? _task.take() : _task.poll(_idleTime, _idleTimeUnit);
+                    // Always poll at some period as safety to ensure we don't poll forever.
+                    Runnable task = _idleTime <= 0 ? _queue.poll(30, TimeUnit.SECONDS) : _queue.poll(_idleTime, _idleTimeUnit);
                     if (LOG.isDebugEnabled())
-                        LOG.debug("{} task={}", this, task);
+                        LOG.debug("{} task={} {}", this, task, ReservedThreadExecutor.this);
                     if (task != null)
                         return task;
 
-                    if (_stack.remove(this))
+                    // Has the RTE stopped?
+                    if (_count.getLo() < 0)
+                        return STOP;
```

Comment:
This should also change the state, like the block below does, so that the caller does not have to differentiate between the 2 cases where it can get back a STOP task.